### ISSUE TITLE
SSO with UI API

### ISSUE THREAD

**Issue by sofdwere** [bgmulinari/B1SLayer#14]
Can you add the ability to login from UI API without username and password, by using _SBO_Application.Company.GetServiceLayerConnectionContext_, like described in [here](https://blogs.sap.com/2016/07/01/sap-business-one-service-layer-sso-with-ui-api/) ?

I have tested it successfully:

example of using:

```C#
string serviceLayerRoot = $"https://{Application.SBO_Application.Company.ServerName}:50000/b1s/v1";
var serviceLayer = new SLConnection(serviceLayerRoot, Application.SBO_Application.Company.GetServiceLayerConnectionContext, 30);
var bpList = await serviceLayer.Request("PriceLists").Filter("PriceListName eq '105900'").Select("PriceListNo").WithCaseInsensitive().GetAsync<List<PriceList>>();
int priceListNo = bpList.First().PriceListNo;
```

code add / changes:

```C#
#region Fields
Func<string, string> GetServiceLayerConnectionContext;



/// <summary>
/// Initializes a new instance of the <see cref="SLConnection"/> class.
/// Only one instance per company/user should be used in the application. 
/// </summary>
/// The Service Layer root URI. The expected format is https://[server]:[port]/b1s/[version]
/// <param name="GetServiceLayerConnectionContext">method from UI api</param>
/// <param name="sessionTimeout">session timeout value of ServiceLayer, see b1s.conf</param>
public SLConnection(string serviceLayerRoot, Func<string, string> GetServiceLayerConnectionContext, int sessionTimeout = 30) {
    ServiceLayerRoot = new Uri(serviceLayerRoot);
    CompanyDB = null;
    UserName = null;
    Password = null;
    Language = null;
    NumberOfAttempts = 3;
    LoginResponse = new SLLoginResponse {
        SessionTimeout = sessionTimeout
    };
    this.GetServiceLayerConnectionContext = GetServiceLayerConnectionContext;
    //
    FlurlHttp.ConfigureClient(ServiceLayerRoot.RemovePath(), client => {
        // Disable SSL certificate verification
        client.Settings.HttpClientFactory = new CustomHttpClientFactory();
        // Ignore null values in JSON
        client.Settings.JsonSerializer = new NewtonsoftJsonSerializer(new JsonSerializerSettings {
            NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore
        });
    });
}



// Session still valid, no need to login again
if (DateTime.Now.Subtract(_lastRequest).TotalMinutes < _loginResponse.SessionTimeout)
    return expectReturn ? LoginResponse : null;
//
if(GetServiceLayerConnectionContext != null) {
    string connectionContext = GetServiceLayerConnectionContext(ServiceLayerRoot.ToString());
    Cookies = new CookieJar();
    string[] cookieItems = connectionContext.Split(';');
    foreach (var cookieItem in cookieItems) {
        string[] parts = cookieItem.Split('=');
        if (parts.Length == 2) {
            Cookies.AddOrReplace(new Flurl.Http.FlurlCookie(parts[0].Trim(), parts[1].Trim(), ServiceLayerRoot.ToString()));
        }
    }
    _loginResponse.LastLogin = DateTime.Now;
    return null;
}
```

I didn't find a way to get the ServiceLayer sessiontimeout value, which is defined in b1s.conf, so for the moment i have added it as parameter with default value (default 30 min). Do you have an idea how it can be done automatically?

**Comment by bgmulinari** [comment]
Hi, @sofdwere. SSO support is something I've considered in the past but haven't got around implementing it. It's definitely on my list of features to add next, so thanks again for your suggestion on this.

**Comment by bgmulinari** [comment]
As for the session control and timeout value, I'll have to look into it as I'm also not sure.

**Comment by bgmulinari** [comment]
Hey, @sofdwere, I just pushed a [commit](https://github.com/bgmulinari/B1SLayer/commit/314df33188975a1c265f8ee31df41ec17a2fd60e) to de dev branch that contains the SSO implementation. I used your code suggestion as a base but changed a few things.

If you are able to download and test it, let me know if it works as intended.

**Comment by sofdwere** [comment]
I have tested the SSO in the current dev branch right now and it works as intended, thank you.

**Comment by bgmulinari** [comment]
Hi, @valkiara. Sorry for the delay on this, I plan to merge this feature soon in an upcoming version.

I will update this issue when the new version is out.

**Comment by valkiara** [comment]
Hi Bruno,

Is there estiamte date when new version will be released?

Regards

**Comment by bgmulinari** [comment]
I apologize again for the delay. This feature is now (finally) available in version [1.3.0](https://github.com/bgmulinari/B1SLayer/releases/tag/1.3.0).

If you encounter any issue, please let me know.

@sofdwere, huge thanks for your suggestion on this.

**Comment by valkiara** [comment]
Thanks a lot Bruno! will let you know if there will be any problem.
